### PR TITLE
Adding SizeMetrics object and constants() method for getting info about problems

### PR DESCRIPTION
**Pull request by nishi951**
Added a SizeMetrics object that serves as an attribute of Problems and contains information about the number of scalar variables, scalar eq and leq constraints, data values, and the maximum dimension of any data block. It is instantiated when the problem is instantiated.

Added a constants() method that functions mostly the same as the variables() and parameters() methods by recursively finding the Constant objects in the objective and constraints of the problem.

### CODE REVIEW

File: cvxpy/atoms/atom.py [hunk 1]
```diff
@@ -209,6 +209,22 @@ def parameters(self):
...
+        """
+        const_list = []
+        const_dict = {}
+        for arg in self.args:
+            const_list += arg.constants()
+        # Remove duplicates:
+        try:
+            const_dict = {str(constant): constant for constant in const_list}
+        except:
+            print const_list
```
Comment (SteveDiamond) [inline cvxpy/atoms/atom.py]: Why is this a try catch?

File: cvxpy/problems/problem.py [hunk 2]
```diff
@@ -143,6 +148,29 @@ def parameters(self):
...
+    def constants(self):
+        """Returns a list of the constants in the problem.
+        """
+        const_dict = {}
+        constants_ = self.objective.constants()
+        for constr in self.constraints:
+            constants_ += constr.constants()
+        # Remove duplicates.
+        # Note that numpy matrices are not hashable, so we use the buildin function id
+        try:
```
Comment (SteveDiamond) [inline cvxpy/problems/problem.py]: Why is this a try catch?

File: cvxpy/problems/problem.py [hunk 3]
```diff
@@ -288,7 +317,7 @@ def _solve(self,
             solver.validate_solver(constraints)
         else:
             raise SolverError("Unknown solver.")
-
+
```
Comment (SteveDiamond) [inline cvxpy/problems/problem.py]: What's up with the spacing here? Is this a tab? No tabs allowed! :)

File: cvxpy/problems/problem.py [hunk 4]
```diff
@@ -574,3 +602,91 @@ def __div__(self, other):
...
+            if constraint.__class__.__name__ is "EqConstraint":
+                self._num_scalar_eq_constr += np.prod(constraint._expr.size)
+        
+        # num_scalar_leq_constr
+        self._num_scalar_leq_constr = 0
+        for constraint in problem.constraints:
+            if constraint.__class__.__name__ is "LeqConstraint":
+                self._num_scalar_leq_constr += np.prod(constraint._expr.size)
+
+    @property
```
Comment (SteveDiamond) [inline cvxpy/problems/problem.py]: You don't need these functions. num_scalar_variables etc can just be raw fields.

**Comment by SteveDiamond** [comment]
Looks good overall. I had a few comments. Also it breaks on python 3. The failure on windows in appveyor also happens for the main cvxpy branch, though if you could figure that one out it would be a huge help.

File: cvxpy/atoms/atom.py (hunk 1 above)
Comment (nishi951) [inline cvxpy/atoms/atom.py]: Whoops, forgot to get rid of all of my debugging stuff. Numpy matrices aren't hashable so I tried to work around that for a while.

Incidentally, it's also written this way in problem.py so I'm changing it there too.

File: cvxpy/problems/problem.py (hunk 2 above)
Comment (nishi951) [inline cvxpy/problems/problem.py]: Same as above, Numpy matrices aren't hashable so I had some errors to debug.

File: cvxpy/problems/problem.py (hunk 3 above)
Comment (nishi951) [inline cvxpy/problems/problem.py]: Ah, sorry my sublime was configured wrong for a while! I must have missed that one.

**Comment by SteveDiamond** [comment]
There's an issue in python 3 on travis. You should test it also in Python 3 locally. This is easy, you just make a new conda environment with Python 3.

**Pull request merged**